### ISSUE TITLE
MLX Phase 2B: Layerwise Observer With Selected Expert Outputs

### ISSUE THREAD

**Issue by egesabanci** [egesabanci/reap-mlx#5]
## Objective

Implement the MLX layerwise observer that replays model layers explicitly, captures selected expert outputs before weighted summation, and emits pruning-compatible observer data.

## Background

The current Torch observers rely on forward hooks and compute all expert outputs with shape roughly `(num_experts, tokens, hidden)`. The MLX implementation should be architecturally different: use explicit replay, exact router adapters, selected expert outputs only, NumPy accumulators, causal masks, and `mx.eval(...)` boundaries.

For pruning, selected expert outputs are sufficient and equivalent to the currently used pruning metrics because REAP/EAN/frequency/max only use outputs from routed experts.

## Scope

Implement observation only. Do not implement pruning mutation, save/reload, CLI, or dataset loader here.

Target files:

- `src/reap/backends/mlx/observer.py`
- tests such as `tests/test_mlx_observer.py`

## Dependencies

This issue depends on:

- MLX Phase 1A router contract and Qwen3 adapter
- MLX Phase 1B NumPy pruning accumulator
- MLX Phase 2A model adapter and MoE layer discovery

## Implementation Steps

1. Add `observe_model(model, tokenizer, calibration_sequences, config, adapter=None)` or a similarly focused API.
2. Accept calibration sequences as unpadded token arrays/lists with batch size 1.
3. Embed tokens using the model's embedding layer through the adapter.
4. For every decoder layer:
   - build or reuse the correct causal mask path
   - run attention through the adapter
   - apply the post-attention normalization
   - if the layer is MoE, call the router adapter
   - call `moe.switch_mlp(moe_input, routing.indices)` or architecture-specific selected expert execution
   - compute selected expert norms/maxes before weighted summation
   - accumulate metrics in `PruningState`
   - compute final MoE output using selected outputs and router scores
   - add shared expert output if present
   - run dense MLP path for non-MoE layers
5. Call `mx.eval(h)` after each layer, or after a documented bounded group of layers.
6. Avoid retaining selected outputs or routing arrays longer than needed.
7. Return `observer_data` keyed by layer index, with the pruning schema from the active plan.
8. Add optional lightweight memory logging behind a debug flag.

## Acceptance Criteria

- [ ] Observer returns a dict keyed by MoE layer index.
- [ ] Each layer report includes `total_tokens`, `expert_frequency`, `pairwise_expert_frequency`, `expert_proba`, `ean_sum`, `ean_mean`, `weighted_ean_sum`, `weighted_expert_frequency_sum`, `reap`, and `max_activations`.
- [ ] `total_tokens` equals the number of real input tokens processed per layer.
- [ ] `expert_frequency.sum()` may exceed `total_tokens` when `top_k > 1` and is tested accordingly.
- [ ] Observer uses causal masks for sequence length greater than 1.
- [ ] Observer captures selected expert outputs before top-k weighted summation.
- [ ] Observer calls `mx.eval(...)` at controlled boundaries.
- [ ] Dense layers in partial-MoE models are forwarded without producing observer state.
- [ ] Importing the observer module does not import `torch` or `vllm`.

## Test Plan

Run:

```bash
pytest tests/test_mlx_no_torch_import.py tests/test_mlx_observer.py -v
```

Suggested tests:

- tiny Qwen3-style model with one MoE layer
- partial-MoE synthetic model with dense and MoE layers
- top-k route counting semantics
- selected-output norm accumulation against a manual reference
- causal mask path is exercised for sequence length greater than 1

## Non-Goals

- Do not support padded batch calibration yet.
- Do not implement merging metrics: TTM, CA, routed characteristic activation, or router-logit similarity.
- Do not download real models in default tests.
- Do not implement CLI orchestration.

## References

- `src/reap/observer.py`
- `src/reap/layerwise_observer.py`
- `implementation-plan-draft.md`, P2-002 and GPT-5.5 addendum
- `docs/reap-mlx-implementation-plan-gpt-v0.2.md`, Layer Replay and Selected Expert Batch

---

## Portfolio Objective Alignment - 2026-05-31

This issue is part of the Apple Silicon MLX REAP backend effort. The production/official REAP experimentation workflow remains the original PyTorch/CUDA implementation; this MLX work must not destabilize or replace that path.

The MLX backend goal is adapter-driven support for MoE weights, not a one-model port. Qwen3-specific work is bootstrap/reference-adapter work. The implementation should preserve REAP pruning semantics and observer-data schema compatibility while adapting to MLX constraints such as lazy execution, no PyTorch hooks, MLX-LM module layouts, and save/reload behavior.

For this issue, the observer must consume adapter-provided routing and expert execution semantics, not Qwen-only assumptions, so any supported MoE weight layout can be observed on MLX.
